### PR TITLE
Allow flags to be set with greater flexibility

### PR DESCRIPTION
**Pull request by daveliddell**
Changes to the python binding to allow iree.runtime.flags.parse_flags to take effect at times other than before the first time a driver is created. Also includes fixes for bugs exposed during the development of this feature.

-    Added "internal" API functions `create_hal_driver()` and `clear_hal_driver_cache()` to create a driver object independent of the cache, and to clear the cache, respectively
- Added `HalDriver` class implementation functions for the above new API functions.  Refactored class to share as much common code as possible.
- Factored out driver URI processing into its own nested class for easier handling of URI components
-    Fixed dangling pointer bug. In the C layer flags are being kept by reference as string views, requiring the caller to keep the original flag strings (argc, argv) around for as long as the flags are being used. However, the python binding was using a local variable for those strings, letting them go out of scope and causing garbage values later on. The fix is to move the strings to a file scope variable. Flag handling does not appear to be getting used in a multi-threaded environment, as other aspects of flag handling use static variables with no mutex guarding that I could find.
-    Fixed runtime assert in Windows debug build for the improper use of std::vector<>::front() on an empty vector. The code never used the value of front(), as it was guarded by a check for the vector's size, but the assert prevents the debug build from running.

### CODE REVIEW

**Comment by daveliddell** [comment]
Please hold off reviewing; will rework client of this API not to use the driver cache, so will git rid of unnecessary new features

**Comment by daveliddell** [comment]
Snippet of ugly but functional use of this new feature:
```
class vmfbRunner:
    def __init__(self, device, vmfb_path, external_weight_path=None, extra_plugin=None):
        flags = []
        clean_driver = False
        if extra_plugin:
            ireert.flags.parse_flags(f"--executable_plugin={extra_plugin}")
            clean_driver = True
        haldriver = ireert.get_driver(device, clean_driver)
```

File: runtime/bindings/python/hal.cc
Comment (stellaraccident) [inline runtime/bindings/python/hal.cc]: It looks like you aren't actually cleaning anything.

How about changing this signature to just take an `std::optional<py::dict*> driver_cache`. If provided, that will be used/updated. If not, then no caching behavior?

File: runtime/bindings/python/hal.cc
Comment (stellaraccident) [inline runtime/bindings/python/hal.cc:1041]: Thank you.

File: runtime/bindings/python/hal.cc
Comment (stellaraccident) [inline runtime/bindings/python/hal.cc]: You don't need to wrap a bool in an optional. Can just default it to false instead of py::none().

File: runtime/bindings/python/hal.cc
Comment (stellaraccident) [inline runtime/bindings/python/hal.cc]: I think this API has exceeded its carrying capacity. Let's split it into three:

* get_cached_hal_driver()
* create_hal_driver()
* clear_hal_driver_cache()

You could merge the first and second by taking a bool `use_cache = True`. I slightly prefer the three-API form, though because "create" and "get_cached" mean different things.

File: runtime/bindings/python/initialize_module.cc
Comment (stellaraccident) [inline runtime/bindings/python/initialize_module.cc]: To make this actually-actually safe, I think you need to make this a `std::vector<std::vector<std::string>*>`. Then every call to `parse_flags` will create a new vector, add it to the global list and keep the strings forever. Otherwise, we have no way to know with repeated calls that a string from a prior call isn't still in use somewhere.

File: runtime/bindings/python/iree/runtime/system_setup.py
Comment (stellaraccident) [inline runtime/bindings/python/iree/runtime/system_setup.py]: Let's switch from `clean = False` to `use_cache = True`. I'd still like the API to actually reset the cache to be private (as in accessible but slightly hard to get to).

File: runtime/bindings/python/initialize_module.cc
Comment (daveliddell) [reply runtime/bindings/python/initialize_module.cc]: Yes, you're right!  I knew that, but was only going for actually safe, not actually-actually safe!  :-D Having chased these strings down to the bottom, they're done with once the flags have been consumed.  That could change at anytime in the future, of course, so I will implement this fix-fix.

File: runtime/bindings/python/hal.cc
Comment (daveliddell) [reply runtime/bindings/python/hal.cc]: The cleaning is making a "clean" new driver object, which replaces whatever was in the cache.  I guess that wasn't clear enough, and hence is a failure of the comprehensibility of my API.  :-(

The intention, however best to express it, is that I didn't want to destroy all of the cached drivers, or require the driver to be separately managed.

File: runtime/bindings/python/hal.cc
Comment (daveliddell) [reply runtime/bindings/python/hal.cc]: Ah.  I was looking for another example of an optional bool in the API and didn't find one.  I'm a nanobind n00b, so I'll have to figure this one out.

File: runtime/bindings/python/hal.cc
Comment (daveliddell) [reply runtime/bindings/python/hal.cc]: If we have separate `get_cached_hal_driver()` and `create_hal_driver()` functions, are users going to think that they should be doing all their driver creation with `create_hal_driver()` and only call `get_cached_hal_driver()` when they want to fetch one that they know they've already created?  That's what it would mean to me.  But then I'm weird.

**Comment by daveliddell** [comment]
General note: in an offline discussion, we decided to keep the cache, so this PR is still a go.  Making it ready-for-review again

File: runtime/bindings/python/hal.cc
Comment (stellaraccident) [reply runtime/bindings/python/hal.cc]: These are still "internal" APIs in that you get at them via the `iree.runtime._binding.get_cached_hal_driver()` namespace. So they are used as part of the user level APIs or for advanced/nefarious ends. 

Being explicit with the names is just me wanting glancibility when looking at implementations so it is clear what is using shared, global state and what is not.

File: runtime/bindings/python/hal.cc
Comment (stellaraccident) [reply runtime/bindings/python/hal.cc]: Ah. I guess I'm just being paranoid/protective about shared global state, and I did misread the intent and the comment (about making a request to clean implying that you wanted the cache cleaned).

I expect that use cases are going to break down between people who don't want to think about it using process global defaults and people who are more sophisticated just wanting to manage their own destiny without process global state making the resource timeline harder to understand. That's why I was thinking to keep the cached and uncached worlds separated: one uses global state and the other does not... Vs half ignoring and mutating the global state.

File: runtime/bindings/python/hal.cc
Comment (daveliddell) [reply runtime/bindings/python/hal.cc]: @stellaraccident What do you think about `get_hal_driver(uri, use_cache=True)`?  Maybe the concept of creation isn't expressed as strongly as in `create_hal_driver`, but "get" can represent both creation and getting a pre-existing object.  That would avoid the confusion I mentioned above.

Comment (daveliddell) [reply runtime/bindings/python/hal.cc]: I guess my problem with `get_cached_hal_driver` is that it doesn't imply creating anything to me, much less so than `get_hal_driver()`.  I would think that that's a worse problem than not knowing what's using global state.  I don't think my API is ideal either, so if I've failed to convince you, you can make an executive decision and tell me what you'd like, and I'd be fine with it. :-)

While I'm whining, "parse_flags" is a terrible name for this function.  It confused 100% of the developers who were looking at it with me.  "add_flags" would have been perfect.

Comment (daveliddell) [reply runtime/bindings/python/hal.cc]: Oh, wasn't thinking about the fact that `get_cached_hal_driver` is already published, so there's nothing that can be done about that unfortunate name.  In that case, I fully agree with you, and I will make three separate functions.

**Comment by daveliddell** [comment]
New sample usage:
```
class vmfbRunner:
    def __init__(self, device, vmfb_path, external_weight_path=None, extra_plugin=None):
        flags = []

        # If an extra plugin is requested, add a global flag to load the plugin
        # and create the driver using the non-caching creation function, as
        # the caching creation function may ignore the flag.
        if extra_plugin:
            ireert.flags.parse_flags(f"--executable_plugin={extra_plugin}")
            haldriver = create_hal_driver(device)

        # No plugin requested: create the driver with the caching create
        # function.
        else:
            haldriver = ireert.get_driver(device)
```

File: runtime/bindings/python/hal.cc
Comment (daveliddell) [reply runtime/bindings/python/hal.cc]: This discussion is made obsolete by the change to the 3-function API.

File: runtime/bindings/python/hal.cc
Comment (daveliddell) [reply runtime/bindings/python/hal.cc]: This discussion is made obsolete by the change to the 3-function API.

File: runtime/bindings/python/hal.cc
Comment (daveliddell) [reply runtime/bindings/python/hal.cc]: Done.

File: runtime/bindings/python/initialize_module.cc
Comment (daveliddell) [reply runtime/bindings/python/initialize_module.cc]: Done.  Now safely safey-safe.

File: runtime/bindings/python/iree/runtime/system_setup.py
Comment (daveliddell) [reply runtime/bindings/python/iree/runtime/system_setup.py]: This discussion is made obsolete by the change to the 3-function API.

File: runtime/bindings/python/initialize_module.cc
Comment (stellaraccident) [inline runtime/bindings/python/initialize_module.cc]: Needs to contain an allocated vector either by unique_ptr or raw pointer (the former is friendlier to unloading). How you have it, the inner vector will not have pointer stability when the outer vector is resized, so all of the contained strings will also not be pointer stable. Just change the inner vector to a unique_ptr and heap allocate it.

File: runtime/bindings/python/initialize_module.cc
Comment (stellaraccident) [inline runtime/bindings/python/initialize_module.cc]: Make_unique once you change above

**Review by stellaraccident — APPROVED** [review]
Approving but please make the requested changes to the flag vector before landing as it has a subtle bug how it is.

File: runtime/bindings/python/initialize_module.cc
Comment (daveliddell) [reply runtime/bindings/python/initialize_module.cc]: Oh, yes, this is true.  Will fix

Comment (daveliddell) [reply runtime/bindings/python/initialize_module.cc]: Probably not getting enough sleep, so very glad to have you watching my back!  Now that I've thought about it some more, `std::vector<std::vector<std::string>>` was probably safe enough.  When the outer vector's data array gets expanded, the inner vectors should be getting move constructed, preserving their own data arrays.  Anyway, reimplemented using `unique_ptr`.

**Comment by ScottTodd** [comment]
CI failure is preexisting and unrelated to this change, merging through it.
https://github.com/iree-org/iree/actions/runs/9512652449/job/26221447494?pr=17659
```
[ RUN      ] NCCLDynamicSymbolsTest.CreateFromSystemLoader
iree/runtime/src/iree/hal/drivers/hip/dynamic_symbols_test.cc:92: Failure
Expected equality of these values:
  21803
  nccl_version
    Which is: 21806
```

**Pull request merged**
